### ISSUE TITLE
Reclaiming of ether in common classes of stuck accounts

### ISSUE THREAD

**Issue by vbuterin** [ethereum/EIPs#156]
### Specification (v1)

The below is only usable between blocks `FORK_BLKNUM` and `FORK_BLKNUM + EFFECTIVE_DURATION` (both TBA).

If the `v` value of a signature is (strictly) greater than 1024, then calculate the sender as follows:
1. Let `s` be the sender computed according to the normal algorithm, using 27 as the `v` value if the provided `v` is odd, and 28 if the provided `v` is even.
2. Let the actual sender be the address that a contract would be created at if its sender is `s` and the contract creation nonce is `floor((v - 1025) / 2)`.

Transactions with `v` values strictly greater than 1024 are only valid if the sender account is nonexistent or its code is empty.

If the `v` value of a signature is 1023 or 1024, then calculate the sender as follows:
1. Let `P` be the **public key** computed according to the normal algorithm, in the 64-byte packed form that is normally hashed to determine the sender address, using 27 as the `v` value if the provided `v` is odd, and 28 if the provided `v` is even.
2. Let the actual sender be the last 20 bytes of the sha3 of the lowercase non-prefixed hex encoded form of `P` instead of the binary raw `P` itself.

### Specification (v2)

Create a solidity contract with the following functions:

* `declareEmptyContract(index)`: computes `rlp.encode([msg.sender, index])`; if there is a contract at this address with ether and no code, then the contract saves a record that this contract has been checked, and sends an equal amount of "future ether" (an ERC20 token) at that account.
* `declareLowercaseHexAddress(pubkey)`: checks `sha3(pubkey) % 2**160 == msg.sender`; then checks that `sha3(pubkey.encode('hex')) % 2**160` has ether; if it does, then the contract saves a record that this contract has been checked, and sends an equal amount of "future ether" (an ERC20 token) at that account
* `withdraw()`: deletes the `msg.sender`'s future ether, and sends it an equivalent amount of ether.

The hard fork would increase the future ether contract's balance by an amount equal to the total quantity of extant future ether.

### Specification (v3)

Follow v2, but distribute the future ether on a Casper testnet. Then, later have a single step that converts both Casper ether and ethereum 1.0 ether into ether as part of the Serenity hardfork.

### Rationale

This allows for users with ether or other assets in common classes of "stuck" accounts to withdraw their assets. The first case covers contracts that are accidentally created with no code, as well as some losses due to replay attacks where a contract was created on ETC, funds sent on ETH but the contract not created on ETH; the second case covers losses due to an old ethereum javascript library that incorrectly computed ethereum addresses. Note that in all cases, the "rightful owner" of the assets is obvious and mathematically provable, and no user is being deprived of any assets, and this proposal provides no explicit favor to any single account, user or application.

It is understood that there may be a risk that this proposal will be viewed controversially as it is in some sense a "rescue" rather than a "technical improvement", even though it is arguably much less intrusive than previous such proposals for the reasons outlined above; the proposal is created in order to allow community discussion and debate and does not signify full endorsement.

**Comment by jespow** [comment]
Speaking on behalf of Kraken, I would characterize this more as recompense than rescue.  We did take on some not insignificant losses as a result of the mentioned bug in the old Ethereum javascript library.  At the time, losses were covered out of Kraken's pocket to protect our clients.  We would greatly appreciate the ability to recover the funds tied up in the incorrectly computed addresses.  Thank you for this proposal.

**Comment by trapp** [comment]
Here are some reports of affected users and implementations:

* https://www.reddit.com/r/ethereum/comments/48rt6n/using_myetherwalletcom_just_burned_me_for/
* https://www.reddit.com/r/ethereum/comments/47nkoi/psa_check_your_ethaddressorg_wallets_and_any/
* https://www.reddit.com/r/ethereum/comments/3vfaob/bug_in_ethereumjsutil/

**Comment by Smithgift** [comment]
Back when the DAO hardfork was debated, this category was brought up as a thing too difficult to fix therefore (INSERT CONCLUSION HERE). I'm glad we're considering fixing it now.

**Comment by cdetrio** [comment]
@jespow @trapp If this spec were to be adopted so that Kraken's losses could be recompensed, would Kraken be willing to in turn cover the losses of users affected by the [not safe enough](https://medium.com/@chevdor/safer-version-of-the-replaysafesplit-smart-contract-a29c347e8a7#.wg1hv4kwg) [ReplaySafeSplit](https://medium.com/@timonrapp/how-to-deal-with-the-ethereum-replay-attack-3fd44074a6d8#.gt6ncetp9) contract?

**Comment by trapp** [comment]
@cdetrio I don't think we can compare the two issues with each other.

1) The ethereumjs-lib did have a bug and did not work as intended.
2) ReplaySafeSplit does not have a bug, it works according to the interface. It has two mandatory address parameters, you have to specify both when calling the contract. You cannot omit one. 

When you specify 0x0 as target address then that is a user input error. When your GUI defaults to 0x0 when you leave one of the two empty, then that's an issue with the GUI. I'm not aware of any GUI that actually does this but see a similar issue in Mist: https://github.com/ethereum/mist/issues/1128

The "fixed" version has an additional check for 0x0 as target address which helps with this specific case but still allows other nonsense addresses like 0x1, 0x2 etc. You cannot fix an issue like that on the contract level, the best place is the GUI (like address checksums).

3) I also think it's offtopic here since the ReplaySafeSplit contract was not created or published by Kraken. The bug in this thread does not only affect Kraken, I would even say that the MyEtherWallet users lost a more due this bug compared to Kraken. Kraken just came forward in support for this initiative.

**Comment by neeeeeeext** [comment]
Another thread related to this EIP:
https://www.reddit.com/r/ethereum/comments/5rd4co/ethereum_wallet_sends_to/

**Comment by GriffGreen** [comment]
Maybe I'm missing it in the complexity... but if someone generates an account offline, for let's say a paper wallet, and then sends to it but claims it was an accident. What is to stop that person from doubling their ether if this EIP is implemented?

**Comment by GriffGreen** [comment]
@LefterisJP and @jbaylina figured it out for me!

http://ethereum.stackexchange.com/questions/760/how-is-the-address-of-an-ethereum-contract-computed#761

The contract address is deterministically computed from the address that created it :-), so the paper wallet trick wouldn't work.

**Comment by axelay** [comment]
Posting on behalf of QuadrigaCX, we would like to see an amendment to this proposal in order to reclaim trapped Ether on contracts that have no ability to send.

Not only to rescue some past mistakes but also thinking about providing a safety net for the future. Even though progresses have been made in Solidity to ensure return of the funds, there could still be a chance of Ether becoming trapped.

**Comment by Souptacular** [comment]
Hi @axelay.

> there could still be a chance of Ether becoming trapped
What are the other ways ether can become trapped? Are you meaning functions that are `payable`, but do not have a means to get the ether out of the function?

Also, please reach out to me at hudson@ethereum.org [PGP 028D 1459 247B 9596 if you prefer] so I can verify that you are representing QuadrigaCX officially. There have recently been people in the community that seem to be speaking for QuadrigaCX, but are not actually employed or represent QuadrigaCX in any way.

Thanks!

**Comment by MichaelMaurer** [comment]
I created a tool to scan for the "ether stuck in empty contract" case. It finds about 3500 ETH in almost 300 empty contracts. I'm not sure I understand the edge cases and welcome any review.

https://gist.github.com/MichaelMaurer/e073091446751b08762c8c65ae13a37d

**Comment by netpro2k** [comment]
> we would like to see an amendment to this proposal in order to reclaim trapped Ether on contracts that have no ability to send.
Wouldn't this be problematic for contracts that are intentionally "burning" ether by locking it up? Sure this could be done in the future in other ways (sending to a dead address) but I would be surprised if there didn't already exist contracts for which this would be an issue.

**Comment by holiman** [comment]
> Wouldn't this be problematic for contracts that are intentionally "burning" ether by locking it up? Sure this could be done in the future in other ways (sending to a dead address) but I would be surprised if there didn't already exist contracts for which this would be an issue.

Intentional burns are usually sending to `0x` or some `0xdead` address. Intentionally locking the ether up in a 'dead' contract to which there is actually a key seems very unlikely.

**Comment by Arachnid** [comment]
> Posting on behalf of QuadrigaCX, we would like to see an amendment to this proposal in order to reclaim trapped Ether on contracts that have no ability to send.

Do you have a technical proposal on how that could be achieved?

Devising a fair rule for this seems particularly problematic, since QuadrigaCX didn't even deploy the contract in question.

**Comment by holiman** [comment]
From a technical perspective, I'd like to point out some things. 

- It's technically possible to (on-chain) verify that a contract `c` only ever can send the `CALLVALUE`.
- It's technically possible to (on-chain) verify that such a contract holds balance. 
- Thus, it could be determined that the `balance` is trapped. 

Now, the hard part would be whom to ascribe the tokens. Because it's impossible _on-chain_ to see who sent what when and if that got stuck. So the only thing that could be done  would be to let the account who deployed the contract get the credits. 

I don't see how it could be done in any other way, from a technical perspective.

**Comment by winsvega** [comment]
It looks more like a client's problem not a protocol issue. If you add a way to send ether back that could potentially compromise the whole system and lead to some weird scenarios that we don't see at the moment. 

The scheme like 
`key -> sign transaction -> send     `
is pretty simple.  

But the scheme like 
`if v> ... and signature is like this then if ...  then make the transaction as if it was if ... `
is not simple and it could break the whole concept IMHO 

The warnings to the user should be done on a client side in a form like: 
"Are you sure that you want to send this transaction to that contract.  This would result in ... <more info> "

**Comment by FlaxFlax** [comment]
Just want to say that I'm one the the noobs that where lucky enough to sent ether with an empty "TO". I know, I know :( 

Now all is stuck in this useless empty contract. I blieve that this will happen again and again. In my case it was a bug but it's quite easy to fall for the believe that it is safe to test send to an empty address as it would fail and bounce the transaction.

Anyhow, any means to reclaim it would be deeply appreciated!

**Comment by sardoru** [comment]
Hi All, I share same situation with @FlaxFlax. Accidentally sent to blank To: field... it was 100 Ether .... :(

**Comment by ryancharleston** [comment]
If/when this feature gets implemented into Ethereum, could it be made to apply to all past transaction or can it only be made to apply to future transactions?

**Comment by timkae** [comment]
Is this something that could be used to help me recover ETH sent to an invalid address?

My story here:  https://www.reddit.com/r/ethereum/comments/6rsvak/sent_3_eth_to_an_address_with_typo/?utm_content=title&utm_medium=user&utm_source=reddit

**Comment by Smithgift** [comment]
@timkae: Unfortunately not. There's no simple way to know if sending to an address was intended or a typo, just from the addresses themselves.

**Comment by tjayrush** [comment]
Has there been any discussion about the following aspect of this original proposal? Quoting from initial post:

> It is understood that there may be a risk that this proposal will be viewed controversially
> as it is in some sense a "rescue" rather than a "technical improvement"

Most of the comments here are from people who obviously agree with the proposal because they lost ETH. Are there any downsides to this proposal? Certainly it gives ammunition to people who argue that ETH is less immutable than it should be. I am not taking a side, but I think the political aspects of the proposal should be discussed.

**Comment by holiman** [comment]
@tjayrush My view on that is that the EIP should concern the technical implementation, so that the technical proposal is finished and fleshed out. Once that is done, and a concrete detailed proposal is available, that proposal should be discussed in a broader scope, which includes the larger community. 

So step 1: concrete technical spec, step 2: determine if it's supported by community. 

...just my 5 cents.

**Comment by MicahZoltu** [comment]
@tjayrush This proposal adds code, and therefore VM complexity, technical debt, etc. and only helps out people who have lost something.  In general, I think anything that requires _adding_ code to all nodes (e.g., part of consensus) needs to have a very strong argument for it.  While I can appreciate the desire to get back lost money, I'm not convinced the total value restored to the userbase outweighs the total cost to the ecosystem.  Do we have any idea exactly how much money would be recoverable with this?  Do we have any idea the level of technical complexity required to implement (and maintain forever) the code associated with this change?

**Comment by mjmau** [comment]
@MicahZoltu With regard to your question of how much money, the answer has two parts: how much has already been lost, and how much will be lost in the future. I tried to estimate how much has been lost already to the subset of "sent to new contract with null code" in my previous comment above, which found around 3500 ETH in that category.

See the gist https://gist.github.com/MichaelMaurer/e073091446751b08762c8c65ae13a37d

**Comment by ryancharleston** [comment]
A recent ICO (REXmls) has 6687 ETH (~$2 million) sitting in an address an address that doesn't actually exist (seen here > https://etherscan.io/address/0x03e4b00b607d0980668ca6e50201576b00000000), due to a misspelled address in their original ICO smart contract. This is a project that would definitely benefit from recovering funds and in doing so would benefit Ethereum and the Ethereum ecosystem as a whole.

**Comment by MicahZoltu** [comment]
@GriffGreen Can you explain how you came to the conclusion that the paper wallet trick won't work?  I also must be missing something, but unfortunately the SE answer you linked wasn't new information to me and even knowing that it still seems like the paper wallet trick would work.
1. Account A create's wallet contract offline.
2. Account B sends funds to wallet.
3. Wait for hardfork.
4. Account B receives refund ETH.
5. Account A uploads wallet contract.
6. Account A withdraws original ETH.

**Comment by plutoegg** [comment]
There are several different suggestions in this thread and elsewhere relating to also including recovery of any **funds accidentally sent to the 0x00 address** .
Since this is a relatively common source of "lost" funds, if these were to be included it could be a special case for the `0x00` address, and would need to be applied regardless of whether the cause was:

1. User error leaving `to` address completely blank, and therefore sending to `0x00`.
2. The early bug, later fixed, with leading whitespaces in the `to` string i.e. `to: " 0x19df848c..."` went instead to `0x00` (https://www.reddit.com/r/ethereum/comments/3gwczr/be_careful_it_is_so_easy_to_make_a_mistake_and/)

The confusion comes that the 0x00 may have been used in the past as a purposeful "burn" address. Therefore any transfers to 0x00 *from contracts* should be treated as having been purposeful attempts to burn ether and be untouched. There is no case where an externally owned account would likely have purposely sent a transfer to `0x00`, and so transfers from these could be treated as "recoverable".

**Comment by DonaldMcIntyre** [comment]
This EIP 156 is a legitimate solution for lost funds as long as it is a generic tool for correcting bonafide mistakes by contract and account owners with lost funds that do not affect other accounts or contracts, or agreements with other unwilling counterparts, even if they are minorities.

As an analogy, if a Bitcoin TX is sent with a very low fee and is never picked up by a miner, it is legit for the owner to withdraw it, raise the fee and re-enter it. Once a TX is confirmed then it cannot be reversed. The only solution is for the receiving account to send back the money, using its private key, and by paying a new fee.

If we regard Ethereum as a global decentralized computing machine, it is a more complex situation. However, any fixes or upgrades to make security, programing, performance, scalability, uptime, reliability and keep integrity of balances and agreements are good.

What must be avoided at all cost is when the distributed computer is used for management of property and agreements between parts. Only the rules the parts deployed must apply. No non-agreed backdoors for any of the parties to unilaterally solve issues. 

Any arbitrary backward reversal of transactions, balances and contract code must be avoided. Any tools that may pose a risk or a backdoor to accounts, balances and contracts must be avoided. 

Another analogy is when people use addresses with wrong format (e.g. a bitcoin format instead of an Ethereum format) they are usually rejected by the system even before they are introduced. This is a legit tool. If EIP 156 has a similar effect then it's legit, IMO.

**Comment by tinybike** [comment]
@GriffGreen I would also like to know why the paper wallet trick won't work; @MicahZoltu's sequence of events looks correct to me.

I share @MicahZoltu's concerns re: complexity risks of this EIP.  Can anyone comment on how much extra code / complexity would be needed to implement this?  My gut feeling is that unless it's _very_ simple, the risk associated with implementation outweighs the benefits of reclaiming lost funds.

**Comment by k26dr** [comment]
Is this being considered as a solution to the bricked Parity multisig wallets?

**Comment by greggdourgarian** [comment]
Can we please be upright and magnanimous and talk straight about what Ethereum is?  Is it immutable and uncensorable as Ethereum.org suggests?  Obviously not, and I'm okay with that given our youthfulness.

Sadly, pull requests to correct to correct it get rejected by Foundation memebers overcome with shame about the DAO hardfork (https://github.com/ethereum/ethereum-org/pull/617#discussion_r125338878).

Myself, I can't stay in a community that doesn't act ethically.   If you want to keep on hard forking to recover funds for people, I really don't care.   But let's at least say what we do and do what we say.

**Comment by tinybike** [comment]
@greggdourgarian while I think your point is worth discussing, IMO this EIP is the wrong place for that discussion.

**Comment by greggdourgarian** [comment]
@tinybike Thanks for responding.    I like this EIP and believe it technically excellent.   

Nevertheless, it crosses the barrier of mere technical improvement and risks defining our ethical framework or lack thereof.   

Given its importance, let's use this opportunity to lift up both Ethereum and the wider community's understanding of blockchain governance.

**Comment by Montoya** [comment]
This seems to me like a technical improvement and a surgical fix to an existing problem. Nothing about this resembles a "hardfork." Naysayers will criticize ETH regardless of what is done, but I would not let their opinions dictate the outcome.

**Comment by Arachnid** [comment]
@Montoya A hard fork is any change to the rules of the chain that requires an update for clients to validate it. This EIP would require a hard fork.

**Comment by greggdourgarian** [comment]
@Montoya You're right - we'll never please everyone.  But we will win over many ideological opponents by articulating and acting on a consistent, ethical framework.    This EIP does represent a hard fork as someone else pointed out, but given the youthfulness of the project, there is nothing inherently wrong with that.

**Comment by supRNurse** [comment]
On July 28th 2016 I tried to use the "safereplaysplitter" contact Vitalik Buterin blogged about in his post "Onward from the hard-fork". It was among my first attempts at using smart contracts and I was very excited. I sent a small amount through per instructions and it worked perfectly. I then sent my entire account through. 1493 Ether never arrived in my account. Instead they were sent by default to 0x0000000000000000000000000000000000000000. Evidently if the drop down menu for the receiving account is deselected or left blank the ether travels by default to the zero address.  My Ether is still there stranded in that un-spendable account. Its loss even at July 2016 prices was devastating. Most people expressed sympathy and they tried to explain "decentralization" without sounding cold. I hold out hope that you developers will find a solution to fix obvious mistakes that doesn't require hard forks. I am so terribly sorry when I hear of good people losing Ether due to unforeseen mistakes. Please guys you are changing the world with Etherium you are our generation's best and brightest please find a solution!
Best
Kevin

**Comment by sciyoshi** [comment]
As noted elsewhere, this EIP as written won't allow withdrawing ether from multisig contracts created using the recently suicided Parity wallet library, since they *do* have code at their address. The class of frozen funds this proposal would allow spending is only those sent to empty/nonexistent contracts, and should be fairly uncontroversial since it gives no preference to user or application.

**Comment by holiman** [comment]
Here's my attempt at a technical implementation for amending the eip: 

Create a solidity contrat with the following function

* `declareSuicidedDependency(address)`: computes `sha3` of the code at `address`. If the code equals `<tbd: code of vulnerable parity-wallet>`, add `futureTokens` corresponding to the `balance` of `address`. 

Notes: Since the address to delegatecall was marked `constant`, it's compiled into the bytescode as a  `PUSH20 0x863df6bfa4469f3ead0be8f9f2aae51c91a907b4`. Had it ben implemented as a dynamic delegatecall operating with storage-variable, this would not have been possible. 

Note2: It is still possible to deposit ether to the contracts, so the `declareSuicideDependency` should  always fully overwrite the `futureTokens` balance with the current `balance`. 

Note3: This solution would not be able to recover tokens. 

Note4: I'm just exploring the technical possibilities here, not championing any particular path forward.

**Comment by zot** [comment]
What about a different type of call transaction aimed at killing contracts with suicided dependencies (something like nukeInvalidContract) that:

1. makes the call
1. if the call succeeds, does an abort (reverting any state changes because this is just a test call)
1. if the call fails because the contract delegates to suicided code, kills the contract
1. gas charges would still apply

Once the contract is dead, the above technique would work to reclaim its ether.

**Comment by shawnholt** [comment]
Not trying to get political, however a thought about a pragmatic solution:

1. There should be a "cost" for this kind of blunder.  Banks charge overdraft fees, this was negligence by one party that affected many other parties.  The negligent party should have to pay (perhaps the Foundation)
2. A clear process for evaluating - POS will help - POW might do the trick.  There is already a voting mechanism.
3. There have to be some clear guidelines - this is a case of frozen ETH, not theft.  There should be a period after which this appeal process expires.  IMO Frozen vs. theft are VERY Different.  The latter is not about immutability, it is about stupidity.

**Comment by flygoing** [comment]
@shawnholt 

1. Enforcing this would be easier said than done, and it would be an entirely political debate as to who is at fault for such issues.

2. In my opinion it is sufficient to allow withdrawals from a contract if A. The requesting address is the address that created the contract, B. The contract contains no CALL/CALLWITH opcodes and therefore could never send ETH from itself, C. The contract address has less than x leading zeros, such that it never collides with reserved addresses used for pre-compiles .

3. Note the first sentence in the original write-up from vitalik, "The below is only usable between blocks FORK_BLKNUM and FORK_BLKNUM + EFFECTIVE_DURATION (both TBA)." This defines a specific block period where these features can be used.

**Comment by shawnholt** [comment]
@flygoing 

Nate - Great point on point 3.  

As for 1. I would still argue that there should be specific guidelines and there should be some governance.  We are talking about an edge case of a maturing technology.  We have a built-in mechanism for voting.  Adding this as a general case without some governance by the community could lead to unintended consequences.  I realize that voting can too but we are talking about a very specific case of frozen assets (as opposed to reversing a transaction or something like that.)  Also, "who is at fault" can defined easily.  Who will pay is where the politics are.  If the developer who killed it is at fault, but Parity wants to pay the fee to protect their brand the politics is outside the protocol. I think using the DAO this way or POS/POW is a great use of self governance - remember these are edge cases.

**Comment by jbaylina** [comment]
For me, the way to proceed with this EIP is:
1.- Make the accountability and determine the place to deposit recovered ETH. This depends very much on each individual case.  The main condition that must meet is:  Cryptographically proved that the ETH is burn and that the deposit belonged to the owner before the burn.
2.- Once the accountability is done and audited many times, we generate the token.
3.- We then make this token transferable. (It can act as a future market of this EIP) And probably it will be tradable at a lower price of ETH
4.- And finally, if there is a consensus in the community, we use a HF to convert the token to real ETH. (May be with a penalty).

My question is: Should we generate a single token for all the cases? Or should we generate different tokens for each case?

In any case, I think it's worthy for affected people to start working on points 1-3 before thinking in point 4.

**Comment by flygoing** [comment]
@shawnholt 

I definitely agree that consensus and governance are important in situations such as this. Ideally a ETH weighted vote similar to that done by Carbonvote during the DAO disaster to decide on the general sentiment in the community, and then, based on the popular vote, the Foundation can react accordingly and act on the most common interest in the community.

Note: The original proposal of this EIP and my amendment in my 2nd point above are indifferent to the cause of the loss of funds, they both just list ways to tell if funds are frozen and how to identify who the true owner should be. While yes I expanded the original criteria to include the recent Parity bug, it isn't directly targeting the people that were affected by that, just any contracts with frozen funds due to there being no CALL opcode.

**Comment by rkagerer** [comment]
I'm new here so take this comment with a grain of salt, but I was really surprised to learn Solidity even lets you make transactions that lock out / burn ETH (aside from some very explicit and intentional opcode).  I feel like generations of accountants learned that lesson a long time ago right before they invented double-entry bookkeeping. :-)

**Comment by Montoya** [comment]
@jbaylina's proposal sounds overly complicated. If the solution to this flaw can be fixed with an upgrade, why do anything else?

**Comment by tjayrush** [comment]
If the community decides to allow voting related to this issue, that voting needs to be presented in a very unbiased way. Humans read from left to right, top to bottom. Any voting instructions or voting results need to be randomized to show each side of the choice either first/last, left/right, top/bottom. The language for each option and all the results must be 100% neutral. And, very importantly, the vote results need to be presented as both by-ether and by-account. In other words, it is equally important how many people voted one way or another as it is how many 'ether' voted one way or another.

To the extent that there is perceived bias in the voting mechanism this already unfortunate event will turn into a disaster. Please be very careful and thoughtful as you proceed.

Also, I think every comment in this thread should end with full disclosure: I have [no|a lot of] ether in any locked in a frozen account.

Full disclosure: I have no ether in any locked accounts.

**Comment by matrixator** [comment]
I realize this might not be the most popular idea, but if the community is going to vote for a HF to unstuck ETH lost by the last Parity bug, shouldn't all problems caused by Parity's bad code be solved? 

I'm thinking about the first multisig hack, that affected so many legit projects... If there's a way, I think it should be done. If another small chain survives because of this, no problem...one more altcoin in this world...it will just make ETC even more insignifiant.

I talked with people from some of the affected communities and even they are somewhat against this idea, so I'm prepared to see the worst comments. Go ahead...shoot. :-)

**Comment by holiman** [comment]
I'd personally prefer to have this EIP discussion be about " Reclaiming of ether in common classes of stuck accounts", as the title says. So discuss the technical suggestions about how stuck ether can be identified, attributed and eventually recovered/reissued. 

The discussion about whether this is a good idea, or if it's even something that the community wants (which is a totally warranted discussion) should imo take place somewhere else.

Full disclosure: I have no ether in any locked accounts.

**Comment by matrixator** [comment]
understood

**Comment by Smithgift** [comment]
On the subject of easy ways to fix this, would it not be simplest to use an irregular state transition to recreate the now-destroyed library, minus the self-destruct? It seems to be to be the simplest method of returning to status quo ante accidentum, at least for this specific incident.

Full disclosure: I have no ether in any locked accounts, or any other lost accounts of a type listed in this EIP.

**Comment by holiman** [comment]
> On the subject of easy ways to fix this, would it not be simplest to use an irregular state transition to recreate the now-destroyed library, minus the self-destruct? It seems to be to be the simplest method of returning to status quo ante accidentum, at least for this specific incident.

If by 'this' you mean the dead parity wallet library, then I agree that technically that is a simpler solution, which also would handle tokens stuck in the wallets. This particular EIP is about a wider class of problems, though.

**Comment by shawnholt** [comment]
@holiman I understand and am not trying to digress into a debate in github, however, I would appreciate a good forum where I can contribute these ideas.  reddit is unfortunately marred with way too much politics and noise, and my points are specific to this EIP.  I'm commenting to help define some requirements for a solution - that **avoids** politics but I do not have sufficient skills to define the technical requirements.  For some projects this is appropriate on github issues.  I'm open to another effective forum and sorry if this is distracting.

The main point is that this is a rare use case and the fix may enable abuse. I believe this is a good example where there should be some governance via voting to avoid abuse and unintended consequences (at least until this is more mature) also to deter usage and compensate the community for a "rescue" there should be a financial penalty (fee) to compensate efforts (e.g. sent to foundation or min gas) and to deter abuse. 

Full disclosure: I have no ether in any locked accounts, or any other lost accounts of a type listed in this EIP.

**Comment by supRNurse** [comment]
Thank you for the discussion. I certainly appreciate your efforts in regards to Ethereum in general and EIP#156 in particular. 

Full Disclosure: I have a ton of Ether stranded in the zero address.

**Comment by 3esmit** [comment]
There is also this contract with 1000 ETH stuck on it https://etherscan.io/address/0x9fa8fa61a10ff892e4ebceb7f4e0fc684c2ce0a9
I've analyzed the problem of this contract and tried to find a solution, managed out that I've could not find a way to retrieve ether from this contract.
The problem there consists that the crowdsale was perfomed by people that + donated ETH + donated effort. The problem is the donated effort, I'll not explain in details here, but it counted the free points when withdrawing (but not give ETH to them) and there was a point where all was witdrawn in storage memory but there was ether left there. 

About parity tech multisig, I agree that if this EIP is activated, should include Parity multisig, which I see one way to fix it is to add fixed code to the library address.

**Comment by footloosejava** [comment]
Not sure if this has been covered already ... If  a contract, not a user, were to send ETH to a contract that becomes terminated, does the ETH go back to the originating contract or the users that sent ETH to the originating contract? And then what if there is no mechanism for users to refund the ETH from the originating contract?

**Comment by jaekwon** [comment]
A few Cosmos fundraiser participants had accidentally sent their would-be contributions to their Cosmos address, which is a Bitcoin HD derived address.  Here's a console tool to generate one for yourself.

```
# First, download from https://github.com/cosmos/fundraiser-cli/releases
jae$ ./cosfundtool_darwin_amd64  --generate

  .d8888b.   .d88888b.   .d8888b.  888b     d888  .d88888b.   .d8888b.
 d88P  Y88b d88P" "Y88b d88P  Y88b 8888b   d8888 d88P" "Y88b d88P  Y88b
 888    888 888     888 Y88b.      88888b.d88888 888     888 Y88b.
 888        888     888  "Y888b.   888Y88888P888 888     888  "Y888b.
 888        888     888     "Y88b. 888 Y888P 888 888     888     "Y88b.
 888    888 888     888       "888 888  Y8P  888 888     888       "888
 Y88b  d88P Y88b. .d88P Y88b  d88P 888   "   888 Y88b. .d88P Y88b  d88P
  "Y8888P"   "Y88888P"   "Y8888P"  888       888  "Y88888P"   "Y8888P
	
Welcome to the Cosmos fundraiser tool.
(Please remember, NEVER type your 12 words onto an "online" computer.)

!!!WARNING!!! Do NOT forget these 12 words.
Your 12-word mnemonic: next ocean trick aim filter narrow transfer legal muscle label bright wolf
!!!WARNING!!! Do NOT forget these 12 words.

!!!WARNING!!! Do NOT use it as an Ethereum address.
Your Cosmos Address: 0x4FB8FAB66EF33955F10D59B9B86183148858DBCE
!!!WARNING!!! Do NOT use it as an Ethereum address.

...
```

The address 0x4FB8FAB66EF33955F10D59B9B86183148858DBCE is not usable as an Ethereum address, but it is the byte-representation of a Bitcoin HD address, e.g. `RipeMD160(Sha256(Secp256k1_PubKey))`.

To put a bounty to this, we will post $1000 worth of ETH today to the above private key.

The largest such transaction was this one: https://etherscan.io/tx/0x086f43fbe16e272468d5f4ced54d8f47c03b45c6c500d51361d6d772917e93d3
2,725 ETH locked up behind a Bitcoin addr.

(BTW, the way to check if you might affected by this issue is to check your Cosmos balance on https://fundraiser.cosmos.network/ . If you lost your Cosmos address but have your seed, then use the `cosfundtool` in the above link.  ALWAYS USE AN OFFLINE AIRGAPPED COMPUTER. If you have issues, please contact us at fundraiser@cosmos.network.)

**Comment by gregrebholz** [comment]
@jaekwon Unless I misunderstand, this EIP would not address the situation you've described.  The two addresses you've shown are "valid" ethereum accounts, it's just that no one apparently possesses the matching private key to control them.  There would be no way for anyone to validate that these transactions were mistakes, and not just transfers to new wallets for which someone did have the matching private key.   The situation described in this EIP is recovering ether sent to 0x00, or, potentially, to deployed contracts with ether but no code (selfdestruct).

Full disclosure: I have no ether in any locked accounts, or any other lost accounts of a type listed in this EIP.

**Comment by holiman** [comment]
@gregrebholz  not quite true:
> The first case covers contracts that are accidentally created with no code, as well as some losses due to replay attacks where a contract was created on ETC, funds sent on ETH but the contract not created on ETH; the second case covers losses due to an old ethereum javascript library that incorrectly computed ethereum addresses.

@jaekwon I think that sounds like it _maybe_ could be covered - but that would assume that it's possible to go from an ethereum address and somehow calculate the corresponding cosmos address. Is that possible? Because the other route would be to submit a private key to the contract via some commit/reveal scheme, which could be messy. 

However, it sounds like the sufficient pieces would be  `ripemd(eth address .. 12 bytes of data)`?

**Comment by ebuchman** [comment]
It would not be possible to go from ethereum address to cosmos address, but it would be possible to go from ethereum public key to cosmos address, since they both use the secp256k1 pubkey. Since we can show that the address in question is the ripemd160(sha256(secp256k1_pubkey)), we can be confident no one has a corresponding private key that could actually unlock it.

**Comment by holiman** [comment]
Sorry, why not? We have the last 20 bytes of keccak (pubkey) in eth address, no?

**Comment by ebuchman** [comment]
Yes, that's right, but we'd have to get the pubkey itself into the EVM. To be clear, the problem is as follows:

- We have `privkey` and `pubkey`
- There are funds in account with `lockedAddress = ripemd160(sha2(pubkey))`. These funds are unspendable since there is obviously no known `privkey2` such that `lockedAddress = keccak(privkey2.pubkey)[12:]`
- We could send a transaction from account `keccak(pubkey)[12:]`
- If a contract can get access to the pubkey of the `msg.sender`, then we could imagine an EIP156 contract could check that `ripemd160(sha2(msg.sender.pubkey) = lockedAddress` and thereby redeem the locked funds to the `msg.sender`. That would be a problem solved! Perhaps this is now possible in Byzantium.

Thanks.

**Comment by shawnholt** [comment]
@jaekwon and others are bringing some variation of use cases.  The EIP refers to a "Stuck Account".  Would be helpful if @vbuterin could make the definition of stuck account a bit more crisp. 

** IMO we should be very careful that this is not in any way a reversal of a valid transaction (ie. DAO use case.) **

Currently lists:
- The first case covers contracts that are accidentally created with no code, as well as some losses due to replay attacks where a contract was created on ETC, funds sent on ETH but the contract not created on ETH
- The second case covers losses due to an old ethereum javascript library that incorrectly computed ethereum addresses.

**Comment by matrixator** [comment]
It's a reversal of state of the network, however you look at it,  but at this early stage it's OK...as kong as we fix ALL problems caused by Parity's numerous bugs.
If we decide it's not worth the pain, because we fear that BTC maximalists will laugh because ETH is just an Excel sheet, then leave it as it is. 
I hope the community has the guts, the strength and sef confidence to do what is right.

**Comment by Montoya** [comment]
> If we decide it's not worth the pain, because we fear that BTC maximalists will laugh because ETH is just an Excel sheet, then leave it as it is.

They have felt this way for a long time, and yet the value of ETH continues to go up. I wouldn't base the decision here on what a clearly biased, defensive group of "competitors" thinks.

**Comment by shawnholt** [comment]
This is a good article on [immutability](https://www.coindesk.com/blockchain-immutability-myth/) - you can skim it as its a long read.

The point I want to reiterate is that we never rollback a valid transaction (even for theft.) Creating a function to recover stuck ETH IMO is totally different and reasonable.  As I mentioned though, with great power comes great responsibility, so I recommend considering a cost and approval process for this scenario.  Economic incentives and oversight should help mitigate malicious, yet unforeseen hacking vectors - at some point the vote could be removed but having a cost could discourage bad coding.

I hope the DAO was the first and last transaction rollback and we can support this EIP without undermining confidence in Ethereum.

**Comment by Arachnid** [comment]
> I hope the DAO was the first and last transaction rollback and we can support this EIP without undermining confidence in Ethereum.

The DAO was not a transaction rollback. Please let's a) try and focus on the topic here, and b) Use correct terminology.

**Comment by holiman** [comment]
@jaekwon @ebuchman Seems to be more complicated than I thought initally. From the seed phrase pasted above:
```
Private key, derived on path 44'/118'/0'/0/0:
	0x499c7dce161a8db03bfaa3a1035156250d932ac917b66e9c08c60b07a20a9dae
COS Public Key               : 0x02D7485091B95C158EEEC8ED2E248F7164BE5DF0956583DB3BB99F326390081CF8
COS Public Key (uncompressed): 0x04D7485091B95C158EEEC8ED2E248F7164BE5DF0956583DB3BB99F326390081CF8F334FDF8ACC7E22C69FDDEDD4FD46DBDDA4F6E635D6F5386350E079101FE5EA8
Eth public key               : 0x04D7485091B95C158EEEC8ED2E248F7164BE5DF0956583DB3BB99F326390081CF8F334FDF8ACC7E22C69FDDEDD4FD46DBDDA4F6E635D6F5386350E079101FE5EA8
Keccak256(pubBytes[1:]) 0xE2DE2CFE745900E94311C55B01E485D6B15E85764B1BE37FA71F2D57AF26B459
Keccak256(pubBytes[1:])[12:] 0x01E485D6B15E85764B1BE37FA71F2D57AF26B459
Public eth: 0x01E485D6B15E85764B1BE37FA71F2D57AF26B459
Cos Address: 0x4FB8FAB66EF33955F10D59B9B86183148858DBCE
Sha256(Cosmos Public Key): 0x88E01006EA89D0400A9FCC7224291FC7A231DEE00A03386A62AE8F4717E16CA7
Kec256(Cosmos Public Key): 0xF0F76DE539979DD340B5DD79F3E28A614F153D77F335245C7FA80DD143AEF7FE
Ripe160(Sha256(Cosmos Public Key)): 0x4FB8FAB66EF33955F10D59B9B86183148858DBCE
```
* Cosmos appear to use compressed public keys, whereas ETH does not. 
* Cosmos uses `sha3`, whereas eth uses `keccak256`. 
* Eth ignores the first byte of the public key, whereas cosmos uses all. 

So it's not just a matter of "add the missing 12 bytes of the sha3-hash and then do ripemd". I think e.g. a sha3 written in solidity would be required, for starters.

**Comment by ebuchman** [comment]
Thanks for adding more details and spelling out the process. That all sounds correct except the piece about `sha3`. Where do we use `sha3` ? Also sha256 and ripemd160 are available as native contracts, so we shouldn't need to implement any hash functions in solidity. Am I missing something ?

**Comment by holiman** [comment]
Doh, I was confused. I'll edit it

**Comment by Arachnid** [comment]
A suggestion: Perhaps it would make more sense to write a generic token contract, and implement specific subclasses of this contract for different recovery scenarios, with their own 'proof' functions.

**Comment by iFA88** [comment]
Greetings, I need also some help with 100 stucked ETH on my contract. The ETH is stucked because the contract was badly coded and I can not withdraw my funds from them.

The contract address is: `0x9122e2cfab13d30237ebeef0c0521d64bf0b06dc`
I have verified the source code too: [etherscan](https://etherscan.io/address/0x9122e2cfab13d30237ebeef0c0521d64bf0b06dc#code)
The contract is created by `0x3F2dA093Bb16EB064F8bfA9E30B929D15f8e1c4C`

Prove to that I own the `0x3F2dA093Bb16EB064F8bfA9E30B929D15f8e1c4C` address private key:
msg: `2017.11.18. Please send my 100 ETH`
msgHash: `ed88c66defbcb5fba064613bb027630935c0c63625350a587caada00a4c745ed`
sign: `0x1eb4a55da612f47e905811ae39a0d4b745e60de3b7f33d7fe5e652aa5e612e18e45ef2bfeda96e919772aeb5d4479066187fadfe45d5429da2cc2692832b4f3c1c`

If you doing hardfork please send back the 100 ETH to the `0x3F2dA093Bb16EB064F8bfA9E30B929D15f8e1c4C` address.

Thanks!

**Comment by Dsummers91** [comment]
There was a reddit thread 8 months ago - [Made a 1 ether contract last year to make a token. How do I recover the ether ?](https://www.reddit.com/r/ethereum/comments/5zve4p/made_a_1_ether_contract_last_year_to_make_a_token/)

Though I am not sure which solidity version it was compiled from, it  has the same function/event hashes and the compiled assembly code seems to closely (but not exact) match to [this contract code](http://chriseth.github.io/browser-solidity/?#gist=21935dc37c5bfbe92e5a&version=soljson-latest.js) which came from the blog post - [Ethereum in practice part 1: how to build your own cryptocurrency without touching a line of code](https://blog.ethereum.org/2015/12/03/how-to-build-your-own-cryptocurrency/). Seems like anyone can send funds to the contract but there is no way to get the funds out. 

If we are able to return those funds that would be great

Contract Address: [0xb9550330bff5b4aada6afcbb6732442c74f80fd7](https://etherscan.io/address/0xb9550330bff5b4aada6afcbb6732442c74f80fd7)
Tx Ether was sent: [0x32e1f0e51ed3486d18379a9053c69ca0cc4825209f889c0c6dd4afcfa0d7a086](https://etherscan.io/tx/0x32e1f0e51ed3486d18379a9053c69ca0cc4825209f889c0c6dd4afcfa0d7a086)

Disclaimer: I am unaffected by this, just trying to help

**Comment by 3esmit** [comment]
@Dsummers91 Lol, they can recover when they want, just call function "sell(amount of tokens they have)" in that token contract. https://etherscan.io/token/0xb9550330BfF5b4AADa6AfcBb6732442c74F80Fd7#balances

**Comment by Dsummers91** [comment]
@3esmit I went through the function hashes from the [assembly](https://gist.github.com/Dsummers91/0fa2d8a41b9bdf0a0f54bdef8f24d39b) and only saw these. 

0x06fdde03 - name()
0x313ce567 - decimals()
0x70a08231 - balanceOf()
0x95d89b41 - symbol()
0xa9059cbb - transfer()

The sell(uint256)  - 0xe4849b32 doesnt exist in there, unless i am missing something..

**Comment by gkratoc1** [comment]
Back in June I stumbled upon a previously undiscovered bug (or lack of failsafe) in Metamask for Firefox which permitted me to send 12.5 ETH to 0x0000000... and I have been watching EIP 156 ever since.  I had almost accepted that the money was gone and never coming back before the latest Parity situation locked up so much more ETH than I lost, and I thought this might make this proposal easier to accept.  I agree with everyone who has said that if it is possible to return the funds to people like myself who obviously had no intention of throwing their money into a black hole then it should be done.  I hope there isn't serious consideration to return the large chunk of money to the big entities without taking care of the little guys who got burned along the way.

The transaction from my horrible day
https://etherscan.io/tx/0x92f699698908c348dcbaf6ecdcc0806dc5b6a08dab7f5df085f7ef070b369b5b

My review of Metamask describing the error is the last one on the list
https://addons.mozilla.org/en-US/firefox/addon/ether-metamask/reviews/

**Comment by 3esmit** [comment]
@Dsummers91 That's sad. Seems like lots of people deployed this exact contract and throwed money there. There is one like this with more then 60 ETH https://etherscan.io/find-similiar-contracts?a=0xb9550330BfF5b4AADa6AfcBb6732442c74F80Fd7&lvl=5

**Comment by fivedogit** [comment]
"the proposal is created in order to allow community discussion and debate and does not signify full endorsement"

Just adding my support for this EIP. For cases where there was an obvious mistake made (funds sent to 0x contracts, etc), reversal is philosophically no different than, say, when geth returns an error on the command line instead of preventing a mistake. While the latter prevents the mistake entirely, the former simply reverses it. 

More generally, I don't think changes accompanying hard forks *must* have technical improvement to be valid and implemented. I have no problem reversing objectively identifiable mistakes.

Disclaimer and background: I very moderately opposed the DAO fork. I have 1 ETH stuck in a 0x contract.

**Comment by tjayrush** [comment]
This comment is intended solely as social commentary. I intend to point out the inherent impossibility of drawing a clear, bright line between legit recovery of lost funds and not-so-obviously-legit recovery.

There is, as of this moment, $109,018,399.22 US, "stuck" in the DAO withdrawal contract. The community should figure out which accounts this money belongs to ("all the data is easily available in the blockchain...") and give the money back. It's the right thing to do. If some particular account has lost its key, then (if they can prove it) we can give them their money back later as part of some other hard fork.

If the community "solves" the lost ether problem for people who are technically adept enough to come to this web page to ask for it, then the community owes it to the less technical people (who can't figure out how to get their money out of the DAO) to return their money.

**I don't actually believe what I've written here.** I'm just trying to point out the danger of this type of EIP. This type of EIP won't "improve" Ethereum. The political (non-technical) ramifications may actually ruin Ethereum. 

Disclosure: I have no ether stuck anywhere.

**Comment by Arachnid** [comment]
> I don't actually believe what I've written here. I'm just trying to point out the danger of this type of EIP. This type of EIP won't "improve" Ethereum. The political (non-technical) ramifications may actually ruin Ethereum.

I don't think constructing a straw man is a particularly good way to do that.

**Comment by gzduan** [comment]
If some particular account has lost its key, then (if they can prove it) we can give them their money back later as part of some other hard fork.

**Comment by LefterisJP** [comment]
@gzduan how do you prove that you used to hold the private key for an account?

**Comment by winsvega** [comment]
it is impossible to prove such a thing. 
you could create some foundation that will payout compensations to those who lost founds by accident.

**Comment by gzduan** [comment]
Can add a variable in the account, if it is 1 means  someone have used the private key of  this account , by default, all the variable of new account  is 0, if someone used the private key of the account, this variable is automatically into 1. Otherwise the variable will always be 0, if you accidentally sent many tokens to the account that variable is 0 , the system will be in after a period of time, such as 3 years ,if this account still didn't have the records of using the private key, in other word,the variable is still 0. The system can identify the account as a black hole account and the money will be automatically returned.

The mistake of sending money to the black hole address, which is unavoidable, will continue to happen now, and this is the defect of the Ethereum network,

**Comment by fivedogit** [comment]
Well we've veered off the path from "mathematically provable ether stuck in contracts" into "ether stuck in accounts whose owners lost keys or passwords". (Unfortunately, I'm in this camp for a significant amount.) For the latter, I thought about this kind of "touch requirement" solution, too. i.e. Require a "touch" of accounts to prove they're not dead, then refund (to sender) those that remain untouched. Without asserting any opinion of my own, I can't imagine the community would allow any requirement like that. Longest of shots, IMO. One can hope, though, right?

Also note with regards to all kinds of stuck ether, there could be dependencies in contracts for an address's balance. ("If 0x123.balance = 5, do something, else do something else.") If we start refunding or reverting or whatever, it's possible conditions based on these balances could change. Shit, conditions based on the REFUND-TO account's balances could change. Although extremely rare, one would think, this is potentially dangerous. (So, personally, I'm back on the fence about whether or not to refund mathematically provable contract-stuck ether.)

**Comment by supRNurse** [comment]
Is there a way to isolate or quarantine indefinitely an address? If Ether is replaced from the un-spendable  account (example my old friend the 0x0000...) can the zero address be isolated or quarantined in order to both prevent any future misdirected ETH landing there?

**Comment by winsvega** [comment]
what you could do is to create another smart contract which would do all the logic you want. 
lets say it will be insurance contract that would have a control of your founds. so if you lose key, you could initiate a procedure to claim your founds back. this way you wont need a hardfork. 

then, if you are afraid that you loose you key, you just registrate your wallet at insurance contract / or should I say wallet contract. 

it will have to be a really well written contract, though. as we dont want it to break again.

**Comment by gzduan** [comment]
Can add a variable in the account, if it is 1 means someone have used the private key of this account , by default, all the variable of new account is 0, if someone used the private key of the account, this variable is automatically into 1. Otherwise the variable will always be 0, if you accidentally sent many tokens to the account that variable is 0 , the system will be in after a period of time, such as 3 years ,if this account still didn't have the records of using the private key, in other word,the variable is still 0. The system can identify the account as a black hole account and the money will be automatically returned.
The mistake of sending money to the black hole address, which is unavoidable, will continue to happen now, and this is the defect of the Ethereum network。
This similar to activate an account,if there have the record of using private key on the blockchain, the account will be automatically activated and will not returned the eth after received the eth three years , the use of paper wallet does is restricted, because there must be activated account operation, but this only to new accounts which private key never used by people , maybe paper wallets are not a necessary a way of storing tokens, Cold wallet signature can be a good choice, can use the private key offline and ensure absolute safety.
Put the tokens sent to the a black hole accounts always happen to us, on the one hand block chain technology ensures the property safety of people, but the careless operation is so despair to ethereum users .the proposal can effectively recover the eth that sent to a black hole account , at the same time ,it also punished the careless users, Because they have to wait at least three years to get the money they lost because of their mistakes.

**Comment by winsvega** [comment]
no need to add such a flag.
if there is a transaction in the network from this account then someone has it's private key,
if there is no transaction from this account, then someone still could have it's private key.

**Comment by gzduan** [comment]
It's like you have to bind a phone number when you use Facebook account

**Comment by Muhammad-Altabba** [comment]
What about tokens? Is it to be considered in this EIP, or should I open separate issue? 
A reported case: https://www.reddit.com/r/ethereum/comments/5rd4co/ethereum_wallet_sends_to/

**Comment by Umiiii** [comment]
Is there any progress about this EIP?
I created 20 empty contracts with no `to` field txns, and I hold the private key for contract creator address.
One of them seems like that:
https://etherscan.io/address/0x3d3507ca6c8f71a166122fe75dfb849810387a21

**Comment by jespow** [comment]
ETH2 seems like a good opportunity to handle this. Treat these coins as burned/staked on ETH1, distribute on the other side.

**Comment by jamslevy** [comment]
@jespow Would the addresses and balances need to be hard-coded in client code or is there some other mechanism to implement this?

**Comment by iFA88** [comment]
> ETH2 seems like a good opportunity to handle this. Treat these coins as burned/staked on ETH1, distribute on the other side.

It would be great, i have "just" 100 ether in my own (typo in source) contract :S

**Comment by plutoegg** [comment]
> @jespow Would the addresses and balances need to be hard-coded in client code or is there some other mechanism to implement this?

This could be implemented using a similar mechanism proposed by Vitalik in 2016 at the start of the thread - using an ERC20 token which distributes lostETH to all affected by the various cases of lost ETH.

An alternate ETH2 deposit contract, would then need to be deployed, accepting deposits of lostETH instead of ETH. ETH2 clients therefore would not need to have any hard-coded addresses or balances, but would need to monitor this second deposit contract for valid deposits, as well as the main deposit contract.

**Comment by jamslevy** [comment]
@plutoegg do you think in practical terms the willingness of ETH2 clients to monitor a deposit contract that would allow for lostETH to be staked would be that different than implementing the change in ETH1? A lot of the same concerns would still apply. I think the one thing that could end a hope of this working is the possibility of a contentious fork, aka "ETH2 Classic". 

One of the considerations would be how this deposit contract actually issues lostETH, making sure it can't be abused, etc. 

Is anyone actually working on this deposit contract based approach to this problem? It seems like the window for it would be in the next few months. 


> > @jespow Would the addresses and balances need to be hard-coded in client code or is there some other mechanism to implement this?
> 
> This could be implemented using a similar mechanism proposed by Vitalik in 2016 at the start of the thread - using an ERC20 token which distributes lostETH to all affected by the various cases of lost ETH.
> 
> An alternate ETH2 deposit contract would then need to be deployed, accepting deposits of lostETH instead of ETH. ETH2 clients therefore would not need to have any hard-coded addresses or balances, but would need to monitor this second deposit contract for valid deposits, as well as the main deposit contract.

**Comment by jacponer** [comment]
Am i also trapped in the black hole? I dont have much knowledge about the chain but can anybody explain this:

Hi, today i sendend a order of 1051 sxp to this adress 0x085a36A9d73abcd18417A9Abe3bA4bc3cB3a51BC by accident. Now i can`t reach this wallet. After some research is saw this is a contract wallet of binance. Can u guys help me to get my coins back? This is the orderadress: https://etherscan.io/token/0x8ce9137d39326ad0cd6491fb5cc0cba0e089b6a9?a=0x085a36a9d73abcd18417a9abe3ba4bc3cb3a51bc

With kind regards,

Gerbert Pater

**Comment by holiman** [comment]
> today i sendend a order of 1051 sxp to this adress 0x085a36A9d73abcd18417A9Abe3bA4bc3cB3a51BC by accident. Now i can`t reach this wallet. After some research is saw this is a contract wallet of binance. 

The `sxp` was sent here:https://etherscan.io/address/0x085a36A9d73abcd18417A9Abe3bA4bc3cB3a51BC . I don't know why you think it's a contract-wallet of binance, as far as I can see there's no code deployed, nor has the address ever been used. Which means that yes, it seems the tokens are trapped in a black hole (caveat: I know nothing about that token, and since it's not ether but a token, it's totally up to the token how funds are managed. A token could e.g. have rollbacks or ways to admin-override balances, it's theoretically possible (if unlikely))

**Comment by jacponer** [comment]
> > today i sendend a order of 1051 sxp to this adress 0x085a36A9d73abcd18417A9Abe3bA4bc3cB3a51BC by accident. Now i can`t reach this wallet. After some research is saw this is a contract wallet of binance.
> 
> The `sxp` was sent here:https://etherscan.io/address/0x085a36A9d73abcd18417A9Abe3bA4bc3cB3a51BC . I don't know why you think it's a contract-wallet of binance, as far as I can see there's no code deployed, nor has the address ever been used. Which means that yes, it seems the tokens are trapped in a black hole (caveat: I know nothing about that token, and since it's not ether but a token, it's totally up to the token how funds are managed. A token could e.g. have rollbacks or ways to admin-override balances, it's theoretically possible (if unlikely))

Ai.. this means i lost the coins forever?

**Comment by holiman** [comment]
As far as I can tell, yes

**Comment by jacponer** [comment]
Hmm, that sucks :) But is it not possible to find this wallet somewhere?

**Comment by djoq** [comment]
I am faced with a similar issue on behalf of a customer. I have no idea how this happened, but he had an address he was trying to send to that ended in 0x...b880 (which is an exchange address and he has no access to private key) and the coins ended up going to the exact same address with the exception of the last two chars, and got locked in 0x...b879. 

If he did have the private key, could the private key be computed of the mistaken address at any reduced cost computationally, knowing the private key of another address differing by a few bytes?

**Comment by masaruduy** [comment]
Some time ago, in 2017, I was using Metamask as a newbie with an interface which still really needed work.  I went and decided to create a contract with only 0x in its code and sent 2.5 eth to it. Will this or other EIP help me get back the funds?

Thanks!

**Comment by supRNurse** [comment]
Believe me I pray it would.. but I haven’t seen any traction trying to move it forward. I can’t see how stranded ETH in the zero addresses utilized this way would be a negative to the integrity of the blockchain.

**Comment by masaruduy** [comment]
> Believe me I pray it would.. but I haven’t seen any traction trying to move it forward. I can’t see how stranded ETH in the zero addresses utilized this way would be a negative to the integrity of the blockchain.

When you say zero addresses, do these include empty contracts?

**Comment by masaruduy** [comment]
Also, how do we push it?!

**Comment by tjayrush** [comment]
> Believe me I pray it would.. but I haven’t seen any traction trying to move it forward. I can’t see how stranded ETH in the zero addresses utilized this way would be a negative to the integrity of the blockchain.

Not to get involved in this conversation (I won't), but I just wanted to say that sometimes people purposefully send ETH to the zero address as a way to burn the ETH. This lowers the number of coins in existence and thereby increases the value of each remaining coin.

The other obvious problem with the suggestion of recovering the ETH currently in the zero address is who gets to decide what to do with it?

There's zero traction on this issue because there's zero good directions to move forward.

**Comment by supRNurse** [comment]
Sometimes it may have been sent to the burn address on purpose true.. I guess but I would Bet the accidents and mistakes hugely outweigh any of those incidents. Send it back to originating addresses stake it, clone it, or whatever let the owners of the originating addresses decide what to do with it... if it was intentional then they can burn it again. No matter and  I understand why many were opposed to even the thought of the idea but I am not one that opposes it but I have a large amount stranded in the zero address and it certainly  wasn’t burned intentionally. Hope all is well with everyone still in this thread.

**Comment by supRNurse** [comment]
Unfortunately probably nothing save waiting till something happens that strands tons of ETH and hope a mechanism for recovery develops that the community can live with.
Best
Kevin

**Comment by masaruduy** [comment]
I understand this case and agree.  However, my situation is different as I
am speaking of sending to a contract that was created with 0x as the code,
not a 0x address itself. Here is an example:

https://etherscan.io/tx/0x3457a2272e1cc0f1834116ed1c55e2502cf5d21f229036761637507882f4da6d


-Nathan

**Comment by supRNurse** [comment]
Don’t lose hope. During the early days of this thread I received messages of support from both individuals and developers. Some even names you would easily recognize. Empathy lives in the chain and one day I trust a pathway for fixing obvious mistakes will present. Good luck and know that I wish you the best.
Kevin

**Comment by masaruduy** [comment]
Thanks for your kind words Kevin 🙂

**Comment by rkagerer** [comment]
There are likewise some 17,000 victims of the Quadriga hack who would benefit if a mechanism like the one discussed here could recover the 67,000 ETH that got [locked up](https://www.coindesk.com/ethereum-client-exchange-14-million) in their [splitter contract](https://etherscan.io/address/0x1e143b2588705dfea63a17f2032ca123df995ce0) when the company upgraded from Geth 1.5.3 to 1.5.9 (@axelay alluded to this in his [2015 comment](https://github.com/ethereum/EIPs/issues/156#issuecomment-306312542)).  That kind of amount would have a real impact on the pot available for distribution.

**Comment by Muhammad-Altabba** [comment]
I hope to find a solution for this soon.
Thanks,

**Comment by Lucienest** [comment]
When you're doing this?

**Comment by djoq** [comment]
Let’s get it down. Pools, claims and voting.

**Comment by Luigy-Lemon** [comment]
This issue seems to be trending back up again...
https://dune.com/queries/1941703?d=11

If people don't want to allow this to be reverted is one thing, it can tricky,  but why has there been no fix to th cause of the problem after 7 years? 
Do not allow wallets to send null transactions to the null address with a positive ether value! 
It seems like such a straighforward rule. Does anyone know the real reason why it has not been addressed?

**Comment by Pandapip1** [comment]
This is no longer the correct place to discuss this anymore. I have created a thread on FEM to continue the conversation: https://ethereum-magicians.org/t/reclaiming-of-ether-in-common-classes-of-stuck-accounts/16111
